atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java

```diff
@@ -28,18 +31,36 @@
 import com.palantir.common.streams.KeyedStream;
 import com.palantir.logsafe.Preconditions;
 import com.palantir.logsafe.SafeArg;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
+import javax.annotation.Nonnull;
+import org.immutables.value.Value;
 
 public class ResilientCommitTimestampPutUnlessExistsTable implements PutUnlessExistsTable<Long, Long> {
+    private static final int TOUCH_CACHE_SIZE = 1000;
+
     private final ConsensusForgettingStore store;
     private final TwoPhaseEncodingStrategy encodingStrategy;
     private final Supplier<Boolean> acceptStagingReadsAsCommitted;
 
+    private final LoadingCache<CellInfo, Long> touchCache = CacheBuilder.newBuilder()
+            .maximumSize(TOUCH_CACHE_SIZE)
```

Comment:
What happens if concurrent transaction rate exceeds cache size? Do we fall off a cliff and increase puts on cache miss?

Reply:
We have to read a staging value in the transactions table, then we try to load. While loading is happening all threads that reach this point will be coalesced. Once the value is loaded it is guaranteed that all reads will forever see the committed value. So the only time a cache hit is actually relevant is if we have read a staging value, but were then so slow we don't get coalesced with someone else. If then our result has already been evicted, it's still fine -- we'll try to do a CAS, fail, but verify it's actually committed now and we'll return the value skipping the put